src/main/java/com/google/devtools/build/lib/analysis/config/transitions/ComposingTransitionFactory.java

```diff
@@ -0,0 +1,93 @@
+package com.google.devtools.build.lib.analysis.config.transitions;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.google.devtools.build.lib.analysis.config.transitions.TransitionFactory.TransitionFactoryData;
+
+/**
+ * A transition factory that composes two other transition factories in an ordered sequence.
+ *
+ * <p>Example:
+ *
+ * <pre>
+ *   transitionFactory1: { someSetting = $oldVal + " foo" }
+ *   transitionFactory2: { someSetting = $oldVal + " bar" }
+ *   ComposingTransitionFactory(transitionFactory1, transitionFactory2):
+ *     { someSetting = $oldVal + " foo bar" }
+ * </pre>
+ */
+@AutoValue
+public abstract class ComposingTransitionFactory<T extends TransitionFactoryData>
+    implements TransitionFactory<T> {
+
+  /**
+   * Creates a {@link ComposingTransitionFactory} that applies the given factories in sequence:
+   * {@code fromOptions -> transition1 -> transition2 -> toOptions }.
+   *
+   * <p>Note that this method checks for transition factoriess that cannot be composed, such as if
+   * one of the transitions is {@link NoTransition} or the host transition, and returns an
+   * efficiently composed transition.
+   */
+  public static <T extends TransitionFactoryData> TransitionFactory<T> of(
+      TransitionFactory<T> transitionFactory1, TransitionFactory<T> transitionFactory2) {
+
+    Preconditions.checkNotNull(transitionFactory1);
+    Preconditions.checkNotNull(transitionFactory2);
+
+    if (transitionFactory1.isFinal()) {
+      // Since no other transition can be composed with transitionFactory1, use it directly.
+      return transitionFactory1;
+    } else if (NoTransition.isInstance(transitionFactory1)) {
+      // Since transitionFactory1 causes no changes, use transitionFactory2 directly.
+      return transitionFactory2;
+    }
+
+    if (NoTransition.isInstance(transitionFactory2)) {
+      // Since transitionFactory2 causes no changes, use transitionFactory1 directly.
+      return transitionFactory1;
+    } else if (NullTransition.isInstance(transitionFactory2) || transitionFactory2.isHost()) {
```

Comment:
Similar to earlier, why not just check transitionFactory2.isFinal() here?

Reply:
Fixed in same way.